SDL JavaScript library (#738)
smartdevicelink/sdl_evolution: proposals/NNNN-sdl-js-library.md

**3. Transport layer**

The first release should come with a base/abstract definition of the transport layer. Based on this definition a TCP socket client transport and a WebSocket Server transport described in [Cloud App Transport Adapter](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0158-cloud-app-transport-adapter.md) should be implemented with this proposal.

#### Manager layer

The first managers that should be introduced in future proposals should be a lifecycle manager and the file manager. Other managers (screen manager etc.) are be added in future releases.

This proposal does not include a manager layer. Managers will be introduced via new proposals.

### Development in TypeScript

The development of this library should not be done in pure JavaScript but in TypeScript. The reason is that TypeScript allows to use data types, classes and other useful development features. This will reduce complexity and risk of issues during development. With TypeScript the maintenance of the library will be much easier compared to pure JavaScript.

> TypeScript will still allow JavaScript support. Whenever a new release version is planned, the TypeScript code can be transcompiled to JavaScript. This [TypeScript Playground](https://www.typescriptlang.org/play/index.html) from typescriptlang.org shows how easy TypeScript can be transcompiled into JavaScript. The export can also provide a source map file for Lint and auto complete when developing JavaScript applications e.g. in VS code. https://www.html5rocks.com/en/tutorials/developertools/sourcemaps/

Reviewer: Can you please create a hyperlink for the html5rocks address?

### Promise based development

Just as the other existing libraries, the JavaScript library requires possibilities to perform async operations. JavaScript is very well capable of serving this task especially with using the Promise pattern. 

There are many sources describing the Promise pattern with JavaScript and how it is used already in the native API and other libraries and frameworks.

> https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Promise

Reviewer: Can you please create a hyperlink for this address?

> https://github.com/getify/You-Dont-Know-JS/blob/master/async%20%26%20performance/ch3.md

Reviewer: Can you please create a hyperlink for this address?

### Project management

> The JavaScript library will be open source and stored in a new smartdevicelink repository on GitHub called sdl_javascript. The project maintainers will be responsible of maintaining and reviewing the code and the project.

Reviewer: ```suggestion
The JavaScript library will be open source and stored in a new smartdevicelink repository on GitHub called sdl_javascript. The project maintainers will be responsible for maintaining and reviewing the code and the project.
```

## Potential downsides

> The downsides of having another SDL library is the increase of SDL maintenance. The project maintainers may need to assign persons who's responsible of managing issues, projects and releases, reviewing pull requests. The SDLC has to decide on the benefits of having a JavaScript library compared to the effort maintaining it.

Reviewer: ```suggestion
The downside of having another SDL library is the increase of SDL maintenance. The project maintainers may need to assign individuals who are responsible for managing issues, projects and releases, and reviewing pull requests. The SDLC has to decide on the benefits of having a JavaScript library compared to the effort maintaining it.
```
